69-common-kingslayer/technical_specification.md

```diff
@@ -0,0 +1,51 @@
+# OpenTelemetry in file services (Common Kingslayer)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+## Scope
+### Outline:
+
+Based on some previous exploration work, this epic aims to implement 1) OpenTelemetry instrumentation in the file services and 2) necessary OpenTelemetry functionality inside of hexkit and/or ghga-service-commons.
+In addition, a viable setup of Jaeger and its components needs to be devised and implemented.
+
+### Included/Required:
+
+While OpenTelemetry supports tracing, metrics and logging, this epic will focus on enabling collection of distributed tracing data and neglect the other two aspects.
+
+#### File Services:
+File services can export tracing data using manual or autoinstrumentation or a combination of both.
+Autoinstrumentation works by installing specific libraries that monkeypatch existing functionality to inject information that can be propagated across service boundaries.
+While these are quite quite useful, they should be vetted for what data is attached to the traces and manual instrumentation should be used instead, if sensitive information is included.
+If autoinstrumentation can be used in all/most cases, additional implementation work can be kept to a minimum.
```

Comment:
Autoinstrumentation should be only used if it is well supported, documented and maintained, and manual instrumentation would be too complicated or time-consuming to implement. Otherwise we may run into problems when upgrading to a newer version of Python or the OpenTelemetry SDK.

Reply:
If we add instrumentation to hexkit, then from the view of the services it would be also kind of auto-instrumentation. We should be very clear what we are talking about.

Reply:
I'll take the second point into consideration, though autoinstrumentation should always mean via the libraries in this context. Might need to specify that in a sentence in the docs.